open-craft/terraform-scripts
# [BB-5321] feat: add cloudwatch provisioning script

## Description

This PR adds, terraform module for provisioning cloudwatch alarms in customer environments.
Apart from cloudwatch alarms, SNS and SNS subscriptions are also setup, to receive alerts when alarm thresholds are breached.

## Supporting information

[BB-5321](https://tasks.opencraft.com/browse/BB-5321)

## Testing instructions

1. Install [Terraform](https://www.terraform.io/downloads)
2.  Install aws-cli:
    ```
    pipx install awscli --force
    ```
3. Configure local AWS profile:
    ```
    aws configure --profile test-profile
    ```
    You will have to enter `AWS Access Key ID`, `AWS Secret Access Key`, and `AWS Region`
4. Create a new file called `main.tf`. Add the below script to it :
```
terraform {
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "~> 3.27"
    }
  }

  required_version = ">= 0.14.9"
}

provider "aws" {
  profile = "test-profile"
  region  = <YOUR_AWS_REGION>
}

resource "aws_instance" "app_server" {
  ami           = "ami-0851b76e8b1bce90b"
  instance_type = "t2.micro"

  tags = {
    Name = "TestInstance"
  }
}


resource "aws_instance" "app_server_2" {
  ami           = "ami-0851b76e8b1bce90b"
  instance_type = "t2.micro"

  tags = {
    Name = "TestInstance2"
  }
}

resource "aws_db_instance" "my_rds_instance" {
  allocated_storage    = 10
  engine               = "mysql"
  engine_version       = "5.7"
  instance_class       = "db.t2.micro"
  name                 = "mydb"
  username             = "foo"
  password             = "foobarbaz"
  parameter_group_name = "default.mysql5.7"
  skip_final_snapshot  = true
}



module "cloud_watch" {
  source = "git@github.com:open-craft/terraform-scripts.git//modules/services/cloudwatch?ref=kaustav/add_cloudwatch_setup"
  default_ec2_alarms_enabled = true
  ec2_instances = [aws_instance.app_server.id, aws_instance.app_server_2.id]
  default_rds_alarms_enabled = true
  rds_instances = [aws_db_instance.my_rds_instance.id]
  environment = "test"
  customer_name = "test-org"
}

resource "aws_sns_topic_subscription" "high_priority_email_subscription" {
  topic_arn = module.cloud_watch.high_priority_alert_arn
  protocol  = "email"
  endpoint = <YOUR_EMAIL_ID>
}
```
5. Replace `<YOUR_AWS_REGION>` and `<YOUR_EMAIL_ID>` with appropriate values
6. Run the below terraform commands:
     ```
     terraform init
     terrofarm apply
     ```
7. Click and confirm any subsciption confirmation email sent by AWS to your email id
8. ssh to the newly provisioned EC2 instance
9. Run the below commands:
     ```
     sudo apt install stress
     stress -c 1
     ```
10. Verify you get notification in email for high CPU usage after about 5-10 mins.
11. Run `terraform destroy` to cleanup
